File: app/src/main/java/com/anokmik/tripassistant/databinding/adapter/BinderRecyclerViewAdapter.java
package com.anokmik.tripassistant.databinding.adapter;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;
import java.util.Map;

public final class BinderRecyclerViewAdapter<T, B extends ViewDataBinding> extends RecyclerView.Adapter<BinderRecyclerViewAdapter.BinderViewHolder<T, B>> {

    private final LayoutInflater inflater;
    private final List<T> items;
    private final ViewHolderPresenter<T> presenter;

    public BinderRecyclerViewAdapter(LayoutInflater inflater, List<T> items, ViewHolderPresenter<T> presenter) {
        this.inflater = inflater;
        this.items = items;
        this.presenter = presenter;
    }

    @Override
    public BinderViewHolder<T, B> onCreateViewHolder(ViewGroup parent, int viewType) {
        return new BinderViewHolder(DataBindingUtil.inflate(inflater, presenter.getLayoutId(), parent, false), presenter);
    }

    @Override
    public void onBindViewHolder(BinderViewHolder holder, int position) {
        holder.bindItem(items.get(position));
    }

    @Override
    public int getItemCount() {
        return (items != null) ? items.size() : 0;
    }

    protected static class BinderViewHolder<T, B extends ViewDataBinding> extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener, AdapterPositionProvider {

        private final B binding;

        private final ViewHolderPresenter<T> presenter;

        private T item;

        public BinderViewHolder(B binding, ViewHolderPresenter<T> presenter) {
            super(binding.getRoot());
            this.binding = binding;
            this.presenter = presenter;
            setAdapterPositionProvider(presenter.getAdapterPositionProviderBindingId());
            setItemClickListener(binding.getRoot(), presenter.getItemClickListener());
            setItemLongClickListener(binding.getRoot(), presenter.getItemLongClickListener());
            setMappedVariables();
        }

        private void bindItem(T item) {
            this.item = item;
            binding.setVariable(presenter.getItemBindingId(), item);
            binding.executePendingBindings();
        }

        @Override
        public int getItemPosition() {
            return getAdapterPosition();
        }

        @Override
        public void onClick(View view) {
            presenter.getItemClickListener().onItemClick(item, getItemPosition());
        }

        @Override
        public boolean onLongClick(View view) {
            presenter.getItemLongClickListener().onItemLongClick(item, getItemPosition());
            return true;
        }

        private void setAdapterPositionProvider(int adapterPositionProviderBindingId) {
            if (adapterPositionProviderBindingId > 0) {
                binding.setVariable(adapterPositionProviderBindingId, this);
            }
        }

        private void setItemClickListener(View view, OnItemClickListener<T> onItemClickListener) {
            if (onItemClickListener != null) {
                view.setOnClickListener(this);
            }
        }

        private void setItemLongClickListener(View view, OnItemLongClickListener<T> onItemLongClickListener) {
            if (onItemLongClickListener != null) {
                view.setOnLongClickListener(this);
            }
        }

        private void setMappedVariables() {
            if (!presenter.isVariablesMapEmpty()) {
                for (Map.Entry<Integer, Object> entry : presenter.getVariablesMapEntries()) {
                    binding.setVariable(entry.getKey(), entry.getValue());
                }
            }
        }

    }

}

File: app/src/main/kotlin/com/anokmik/tripassistant/databinding/adapter/BinderRecyclerViewAdapter.kt
package com.anokmik.tripassistant.databinding.adapter

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

class BinderRecyclerViewAdapter<T, B : ViewDataBinding>(private val inflater: LayoutInflater, private val items: List<T>, private val presenter: ViewHolderPresenter<T>) : RecyclerView.Adapter<BinderRecyclerViewAdapter.BinderViewHolder<T, B>>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BinderViewHolder<T, B> = BinderViewHolder(DataBindingUtil.inflate<ViewDataBinding>(inflater, presenter.layoutId, parent, false), presenter)

    override fun onBindViewHolder(holder: BinderViewHolder<T, B>, position: Int) = holder.bindItem(items[position])

    override fun getItemCount() = items.size

    class BinderViewHolder<in T, in B : ViewDataBinding>(private val binding: B, private val presenter: ViewHolderPresenter<T>) : RecyclerView.ViewHolder(binding.root), View.OnClickListener, View.OnLongClickListener {

        private var item: T? = null

        init {
            setItemClickListener(itemView, presenter.itemClickListener)
            setItemLongClickListener(itemView, presenter.itemLongClickListener)
            setMappedVariables()
        }

        fun bindItem(item: T?) {
            this.item = item
            binding.setVariable(presenter.itemBindingId, item)
            binding.executePendingBindings()
        }

        override fun onClick(view: View) {
            presenter.itemClickListener?.onItemClick(item, adapterPosition)
        }

        override fun onLongClick(view: View): Boolean {
            presenter.itemLongClickListener?.onItemLongClick(item, adapterPosition)
            return true
        }

        private fun setItemClickListener(view: View, onItemClickListener: OnItemClickListener<T>?) {
            onItemClickListener?.let {
                view.setOnClickListener(this)
            }
        }

        private fun setItemLongClickListener(view: View, onItemLongClickListener: OnItemLongClickListener<T>?) {
            onItemLongClickListener?.let {
                view.setOnLongClickListener(this)
            }
        }

        private fun setMappedVariables() {
            presenter.variablesMap?.apply {
                if (!isEmpty()) {
                    for ((key, value) in entries) {
                        binding.setVariable(key, value)
                    }
                }
            }
        }

    }

}